### PR TITLE
Simplify createFromSends

### PR DESCRIPTION
**Pull request by aprokop**
We have no reason to do multiple rounds of MPI communication here, all
of them can be done in a single Alltoall.

### CODE REVIEW

**Comment by aprokop** [comment]
This is just something that bothered me after I looked at it.

File: src/details/ArborX_DetailsDistributor.hpp [hunk 1]
```diff
@@ -107,45 +107,29 @@ class Distributor
...
-    // exclusive scan
-    _src_offsets[0] = 0;
-    for (int i = 0; i < indegrees; ++i)
-      _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+    _src_offsets.push_back(0);
+    for (int i = 0; i < comm_size; ++i)
+      if (mask[i])
+      {
+        _sources.push_back(i);
+        _src_counts.push_back(mask[i] - 1);
```
Comment (masterleinad) [inline src/details/ArborX_DetailsDistributor.hpp]: You should also document that this is to restore the correct size after adding 1 above.

File: src/details/ArborX_DetailsDistributor.hpp [hunk 2]
```diff
@@ -107,45 +107,29 @@ class Distributor
...
-        mask[i] = 1;
-      MPI_Alltoall(MPI_IN_PLACE, 0, MPI_DATATYPE_NULL, mask.data(), 1, MPI_INT,
-                   _comm);
-      for (int i = 0; i < comm_size; ++i)
-        if (mask[i] == 1)
-          _sources.push_back(i);
+      // The previous behaviour was that even if _dest_counts[i] == 0, we still
+      // added that to the source lists. Here, we add 1 to retain that, even
+      // though it is weird as of why we would have such cases.
+      mask[_destinations[i]] = _dest_counts[i] + 1;
```
Comment (masterleinad) [inline src/details/ArborX_DetailsDistributor.hpp]: What happens if we don't do it? At first sight, this really looks unnecessary.

Comment (aprokop) [reply src/details/ArborX_DetailsDistributor.hpp]: If we remain in the scope of this function, and don't do that, then the senders are going to be different from receivers, and things are going to deadlock somewhere. I'm not sure about ramifications of that, and whether this happens.

Comment (masterleinad) [reply src/details/ArborX_DetailsDistributor.hpp]: I see. Fine with me then.

**Comment by dalg24** [comment]
Also not sure if you remember how we got to the current implementation.  Originally we we going to go with `MPI_Neighbor_alltoallv` for the actual communication but we were getting random segfaults so I started building the graph by hand and we ended up with `MPI_Isend/MPI_Irecv`

**Comment by aprokop** [comment]
I've updated the PR, removing +-1.

File: src/details/ArborX_DetailsDistributor.hpp [hunk 3]
```diff
@@ -107,45 +107,26 @@ class Distributor
...
-      MPI_Waitall(requests.size(), requests.data(), MPI_STATUSES_IGNORE);
-
-    _src_offsets.resize(indegrees + 1);
-    // exclusive scan
-    _src_offsets[0] = 0;
-    for (int i = 0; i < indegrees; ++i)
-      _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+    _src_offsets.push_back(0);
+    for (int i = 0; i < comm_size; ++i)
+      if (src_counts_dense[i])
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributor.hpp]:
```suggestion
      if (src_counts_dense[i] > 0)
```
Readability.  If I don't complain, I guarantee ClangTidy will.

**Pull request merged**